packages/vite/src/shared/moduleRunnerTransport.ts

```diff
@@ -0,0 +1,292 @@
+import { nanoid } from 'nanoid/non-secure'
+import type { CustomPayload, HotPayload } from 'types/hmrPayload'
+import { promiseWithResolvers } from './utils'
+
+export type ModuleRunnerTransportHandlers = {
+  onMessage: (data: HotPayload) => void
+  onDisconnection: () => void
+}
+
+/**
+ * "send and connect" or "invoke" must be implemented
+ */
+export interface ModuleRunnerTransport {
+  connect?(handlers: ModuleRunnerTransportHandlers): Promise<void> | void
+  disconnect?(): Promise<void> | void
+  send?(data: HotPayload): Promise<void> | void
+  invoke?(
+    data: HotPayload,
+  ): Promise<{ /** result */ r: any } | { /** error */ e: any }>
```

Comment:
Is it possible to have `invoke(data: CustomPayload)`? This gives type error below in `sendOrInvoke`, but I cannot tell if `sendOrInvoke` is necessary there.

Reply:
We can limit it to `CustomPayload` and remove `sendOrInvoke`. But if we do that, we cannot use `send` later on without telling users to implement it.
But ya, I agree we should make `CustomPayload` vs `HotPayload` more clear.

Reply:
> But if we do that, we cannot use `send` later on without telling users to implement it.

On second thought, this doesn't work unless the `this.hot.on` is called, so it doesn't make sense unless we force users to always implement HotChannel...

Reply:
OK. I've made users to always use `HotChannel`. https://github.com/vitejs/vite/pull/18362#discussion_r1816436241 should be solved as well.